server/app/views/questiontypes/AddressQuestionRenderer.java

```diff
@@ -71,11 +75,12 @@ protected DivTag renderTag(
                             .setFieldName(addressQuestion.getCityPath().toString())
                             .setLabelText(messages.at(MessageKey.ADDRESS_LABEL_CITY.getKeyName()))
                             .setValue(addressQuestion.getCityValue().orElse(""))
-                            .addReferenceClass(ReferenceClasses.ADDRESS_CITY)
                             .setFieldErrors(
                                 messages,
                                 validationErrors.getOrDefault(
                                     addressQuestion.getCityPath(), ImmutableSet.of()))
+                            .setAriaInvalid(hasErrors)
```

Comment:
I'm a bit confused here and elsewhere around why `hasErrors` is piped around and why we're setting the aria-invalid attribute here based on a global concept of whether errors exist. Wouldn't this only be invalid if the above argument to `setFieldErrors` was non-empty?

If that's the case, is it possible we could bundle up all of the logic for setting this attribute into FieldWithLabel and transparently set the aria-invalid attribute if the field errors are non-empty? That'd also allow us to remove the `setAriaInvalid` method from the class.

Reply:
I believe top-level question errors are surfaced as an entry in the map at the top-level question's path. We extract that [here](https://github.com/civiform/civiform/blob/main/server/app/views/questiontypes/ApplicantQuestionRendererImpl.java#L82-L97). If a "required" question doesn't end up surfacing an error there, that seems like a bit of debt that we could clean up. If we wanted to separate it from this PR, we could perhaps create a copy of the map with a required error and pass it along to downstream callers so that they could continue to use the map as the "source-of-truth".

> My impression was that validationErrors held a map for validation errors across all questions. Is it actually per question? 

I believe it's a map constructed based on a specific question instance ([code](https://github.com/civiform/civiform/blob/29dd8296eefb3e59aad3ad97cb41fa4d9e6a5ef9/server/app/services/applicant/question/Question.java#L51)).

Re: `setAriaInvalid`, sounds good to me!

Reply:
In that first link you provided, it actually uses [isRequiredButWasSkippedInCurrentProgram()](https://github.com/civiform/civiform/blob/main/server/app/services/applicant/question/ApplicantQuestion.java#L108:18), which uses various properties of the question object instead of the map. I agree though that we should do a cleanup so that it's represented in that error map. Seems confusing that the error logic is fragmented. I created #3393, and will do a fast follow fix.